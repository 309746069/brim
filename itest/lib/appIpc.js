/* @flow */
import {Application} from "spectron"
import type {Space} from "../../src/js/lib/Space"
import Field from "../../src/js/models/Field"
import Log from "../../src/js/models/Log"
import menu from "../../src/js/electron/menu"
import {LOG} from "./log"

// This module exists to work around difficulties in using Spectron to drive
// Brim with its components that use Electron's native support. Right now
// that's the right-click menu, which has recently gone native. See the
// long-standing issue here:
// https://github.com/electron-userland/spectron/issues/21
export default function contextMenuShim(
  app: Application,
  program: string,
  columns: string[],
  space: Space
) {
  return {
    program(value: string) {
      program = value
      return this
    },

    click(label: string, field: Field, log: Log) {
      LOG.debug(
        'Creating menu from log entry on field "' + field.toString() + '"...'
      )
      let ipcMenu = menu.fieldContextMenu(program, columns, space)(
        field.toBrimField(),
        log,
        false
      )
      LOG.debug(`Finding menu item with label "${label}"...`)
      let item = ipcMenu.find((item) => item.label && item.label === label)
      // The item.click test below is needed to make flow happy.
      if (item && item.click) {
        LOG.debug("About to click menu item...")
        item.click(item, {webContents: app.webContents})
      } else {
        throw new Error("Could not find '" + label + "' in context menu")
      }

      return this
    }
  }
}
