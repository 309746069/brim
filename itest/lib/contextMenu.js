/* @flow */
import type {Space} from "../../src/js/lib/Space"
import Field from "../../src/js/models/Field"
import Log from "../../src/js/models/Log"
import menu from "../../src/js/electron/menu"

export default function contextMenu(
  app: *,
  program: string,
  columns: string[],
  space: Space
) {
  return {
    program(value: string) {
      program = value
      return this
    },

    click(label: string, field: Field, log: Log) {
      let item = menu
        .fieldContextMenu(program, columns, space)(
          field.toBrimField(),
          log,
          false
        )
        .find((item) => item.label && item.label === label)

      if (item && item.click) {
        item.click(item, {webContents: app.webContents})
      } else {
        throw new Error("Could not find '" + label + "' in context menu")
      }

      return this
    }
  }
}
