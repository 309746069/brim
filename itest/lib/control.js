/* @flow */

export const retry = (f, attempts = 100, delay = 100) => {
  return new Promise((resolve, reject) => {
    f()
      .then((ret) => resolve(ret))
      .catch((err) => {
        setTimeout(() => {
          if (attempts === 1) {
            reject(err)
          } else {
            retry(f, attempts - 1, delay)
              .then((ret) => resolve(ret))
              .catch((err) => reject(err))
          }
        }, delay)
      })
  })
}

export const retryUntil = (f, cond_f, attempts = 5, delay = 1000) =>
  // Retry f until cond_f is true. Return results from f.
  //
  // f: function that returns a promise
  // cond_f: function that expects f's resolved value and returns a Boolean.
  // If cond_f returns false, then f is retried. This attempts every delay ms
  // for attempts before rejecting.
  retry(
    () =>
      new Promise((resolve, reject) => {
        f()
          .then((val) => {
            if (cond_f(val)) {
              resolve(val)
            } else {
              reject(new Error(`retryUntil condition failure: ${val}`))
            }
          })
          .catch((err) => {
            reject(`retryUntil promise failure: ${err}`)
          })
      }),
    attempts,
    delay
  )
