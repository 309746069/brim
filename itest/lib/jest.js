/* @flow */

import {Application} from "spectron"

import {selectors} from "../../src/js/test/integration"

export const TestTimeout = 60000

// This is trying to make Flow happy for passing in the done function from
// Jest.
export type Done = {
  (string | Error | void): void,
  fail?: (string | Error | void) => void
}

export const handleError = async (
  app: Application,
  initialError: Error,
  done: Done
) => {
  let realError = undefined
  let notificationError = undefined
  console.log(`handleError: Test hit exception: ${initialError.message}`)
  console.log("handleError: Looking for any desktop app notifications")
  try {
    notificationError = await app.client.getHTML(selectors.notification, false)
  } catch (e) {
    notificationError = undefined
  }
  if (notificationError) {
    realError = new Error(
      "App notification '" +
        notificationError +
        "' (initial error: '" +
        initialError +
        "'"
    )
  } else {
    console.log("handleError: desktop app notification not found")
    realError = initialError
  }
  done.fail && done.fail(realError)
}
