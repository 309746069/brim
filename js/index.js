import React from "react"
import ReactDOM from "react-dom"
import {AppContainer} from "react-hot-loader"
import {Provider} from "react-redux"
import initStore from "./initStore"
import {HashRouter} from "react-router-dom"
import XApp from "./connectors/XApp"
const store = initStore()

ReactDOM.render(
  <AppContainer>
    <Provider store={store}>
      <HashRouter>
        <XApp />
      </HashRouter>
    </Provider>
  </AppContainer>,
  document.getElementById("app-root")
)
