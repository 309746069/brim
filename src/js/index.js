import React from "react"
import ReactDOM from "react-dom"
import {Provider} from "react-redux"
import initStore from "./initStore"
import {HashRouter} from "react-router-dom"
import XApp from "./connectors/XApp"
const store = initStore()

ReactDOM.render(
  <Provider store={store}>
    <HashRouter>
      <XApp />
    </HashRouter>
  </Provider>,
  document.getElementById("app-root")
)
