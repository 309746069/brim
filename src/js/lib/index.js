/* @flow */

import file from "./file"
import keep from "./keep"
import obj from "./obj"
import window from "./window"

export default {
  file,
  keep,
  obj,
  window
}
