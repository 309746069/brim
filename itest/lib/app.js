/* @flow */
const electronPath = require("electron")

import {Application} from "spectron"
import * as path from "path"

import {downloadsDir} from "../../src/js/lib/System"
import {selectors} from "../../src/js/test/integration"
import {LOG} from "./log"
import {workspaceLogfile} from "../lib/log"
import {retryUntil} from "./control"

const appStep = async (stepMessage, f) => {
  LOG.debug(`Starting step "${stepMessage}"`)
  let result = await f()
  LOG.debug(`Finished step "${stepMessage}"`)
  return result
}

export const newAppInstance = (name: string, idx: number) =>
  // https://github.com/electron-userland/spectron#new-applicationoptions
  new Application({
    path: electronPath,
    args: [path.join(__dirname, "..", "..")],
    // PROD-853: Spectron will choose a random debugging port in the 9000-9999
    // range. Since boomd uses 9867, set this to 9999.
    chromeDriverArgs: ["remote-debugging-port=9999"],
    startTimeout: 60000,
    waitTimeout: 60000,
    chromeDriverLogPath: workspaceLogfile(
      name + idx.toString() + "-chromedriver.log"
    ),
    webdriverLogPath: workspaceLogfile(name + "-webdriverLogFiles"),
    // PROD-831: Latest compatible spectron and webdriverio lead to the
    // following:
    //  console.warn node_modules/webdriverio/build/lib/helpers/deprecationWarning.js:12
    //    WARNING: the "<cmd>" command will be deprecated soon. If you have further questions, reach out in the WebdriverIO Gitter support channel (https://gitter.im/webdriverio/webdriverio).
    //    Note: This command is not part of the W3C WebDriver spec and won't be supported in future versions of the driver. It is recommended to use the actions command to emulate pointer events.
    //
    //    (You can disable this warning by setting `"deprecationWarnings": false` in your WebdriverIO config)
    // for <cmd> in the following:
    //   buttonPress
    //   moveTo
    // Test code isn't using these directly according to git grep, which
    // means this is something one of my dependencies must fix. Ignore the
    // warnings for now.
    webdriverOptions: {
      deprecationWarnings: false
    }
  })

export const startApp = async (app: Application) => {
  await appStep("starting app", () => app.start())
  return await resetState(app)
}

export const waitForLoginAvailable = (app: Application) => {
  // Wait for login elements to exist. In most cases this is taken care of for
  // you in logIn() however if you need to inspect those elements before
  // logging in, use this first.
  const waitForHostname = () => {
    return app.client.waitForVisible(selectors.login.host)
  }

  const waitForButton = () => {
    return app.client.waitForVisible(selectors.login.button)
  }
  return appStep("wait for hostname, port, and login widget", () =>
    waitForHostname().then(() => waitForButton())
  )
}

export const logIn = (app: Application) => {
  // Wait for necessary login widgets and then log in. Then make sure the app
  // is ostensibly ready before continuing. This method is suitable for most
  // test procedures.
  return waitForLoginAvailable(app)
    .then(() =>
      appStep("fill out login page and log in", () =>
        // WebdriverV4 doesn't return promises for these methods. Instead they can
        // be chained together.
        app.client
          .setValue(selectors.login.host, "localhost:9867")
          .click(selectors.login.button)
      )
    )
    .then(() => waitForSearch(app))
    .then(() => waitForHistogram(app))
}

export const resetState = (app: Application) =>
  appStep("reset state", () => app.webContents.send("resetState"))

export const waitForSearch = (app: Application) => {
  return appStep(
    "wait for main search input to appear and then get its value",
    () =>
      app.client
        .waitForVisible(selectors.search.input)
        .then(() => app.client.element(selectors.search.input).getValue())
  )
}

export const waitForHistogram = (app: Application) => {
  return appStep("wait for base histogram element hierarchy", () =>
    app.client
      .waitForVisible(selectors.histogram.topLevel)
      .then(() =>
        app.client.element(selectors.histogram.topLevel).getAttribute("class")
      )
  )
}

export const writeSearch = (app: Application, searchText: string) =>
  appStep("write to main search", () =>
    app.client
      .waitForVisible(selectors.search.input)
      .then(() => app.client.setValue(selectors.search.input, searchText))
  )

export const getSearchText = (app: Application): Promise<string> =>
  appStep("get contents of main search", () =>
    app.client
      .waitForVisible(selectors.search.input)
      .then(() => app.client.getValue(selectors.search.input))
  )

export const startSearch = (app: Application) =>
  appStep("click the search button", () => app.client.keys("Enter"))

export const searchDisplay = async (app: Application) => {
  // This stinks. We have to use getHTML because headers that are off the
  // screen return as empty strings if you use getText. This isn't required of
  // actual results.
  // See http://v4.webdriver.io/api/property/getText.html
  // app.browserWindow.maximize() fixes the problem on my laptop but not CircleCI.
  // But what we get back includes the width which can be non-deterministic:
  // <div class="header-cell" style="width: 192px;">ts<div class="col-resizer"></div></div>
  // That style width will vary on my laptop vs. CircleCI.
  // The hack is to split this and extract just the text.
  const _trim = (s: string) => s.split(">")[1].split("<")[0]

  const headerResults = () => {
    return appStep("get search fields", () =>
      app.client
        .waitForVisible(selectors.viewer.headers)
        .then(() => app.client.getHTML(selectors.viewer.headers))
    ).then((headers) => {
      if (typeof headers === "string") {
        headers = [headers]
      }
      return headers.map((h) => _trim(h))
    })
  }
  const searchResults = () =>
    appStep("get search tuples", () =>
      app.client
        .waitForVisible(selectors.viewer.results)
        .then(() => app.client.getText(selectors.viewer.results))
    )

  let headers = await headerResults()
  let search = await searchResults()
  // $FlowFixMe
  return headers.concat(search)
}

export const getCurrentSpace = (app: Application) =>
  appStep("get the current space", () =>
    app.client
      .waitForVisible(selectors.spaces.button)
      .then(() => app.client.getText(selectors.spaces.button))
  )

export const setSpace = (app: Application, space: string) =>
  appStep(`set space to "${space}"`, () =>
    app.client
      .waitForVisible(selectors.spaces.button)
      .then(() => app.client.click(selectors.spaces.button))
      .then(() => app.client.waitForVisible(selectors.spaces.menuItem(space)))
      .then(() => app.client.click(selectors.spaces.menuItem(space)))
  )

const getSearchStat = (app: Application, selector: string) =>
  appStep(`get search stats for selector "${selector}"`, () =>
    app.client
      .waitForVisible(selector)
      .then(() => app.client.getText(selector))
      .then((text) => parseFloat(text.split(" ")[0]))
  )

export const getSearchSpeed = (app: Application) =>
  getSearchStat(app, selectors.search.speed)

export const getSearchTime = (app: Application) =>
  getSearchStat(app, selectors.search.time)

export const setSpan = (app: Application, span: string) => {
  const clickSpanButton = () =>
    appStep("click span selector", () =>
      app.client
        .waitForVisible(selectors.span.button)
        .then(() => app.client.click(selectors.span.button))
    )

  const clickSpan = () =>
    appStep(`select span ${span}`, () =>
      app.client
        .waitForVisible(selectors.span.menuItem(span))
        .then(() => app.client.click(selectors.span.menuItem(span)))
    )

  return clickSpanButton().then(() => clickSpan())
}

export const click = (app: Application, selector: string) =>
  appStep(`click on selector "${selector}"`, () =>
    app.client.waitForVisible(selector).then(() => app.client.click(selector))
  )

export const rightClick = (app: Application, selector: string) =>
  appStep(`right-click on selector "${selector}"`, () =>
    app.client
      .waitForVisible(selector)
      .then(() => app.client.rightClick(selector))
  )

export const openDebugQuery = async (app: Application) => {
  await click(app, selectors.options.button)
  await click(app, selectors.options.menuItem("Debug query"))
  await appStep("wait for Debug Query modal to appear", () =>
    Promise.all([
      app.client.waitForVisible(selectors.debugSearch.search),
      app.client.waitForVisible(selectors.debugSearch.ast),
      app.client.waitForVisible(selectors.debugSearch.done)
    ])
  )
}

export const setDebugQuery = (app: Application, searchText: string) =>
  app.client.setValue(selectors.debugSearch.search, searchText)

export const getDebugAst = (app: Application) =>
  app.client
    .getText(selectors.debugSearch.ast)
    .then((astText) => JSON.parse(astText.join("")))

export const waitUntilDownloadFinished = async (app: Application) =>
  await appStep("wait for a download to finish", async () => {
    await app.client.waitForVisible(selectors.downloadMessage)
    return await retryUntil(
      () => app.client.getText(selectors.downloadMessage),
      (text) => text == "Download Complete" || text.includes("Download error")
    )
  })

export const pcapsDir = () => downloadsDir()
