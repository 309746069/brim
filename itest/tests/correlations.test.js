/* @flow */

const electronPath = require("electron")

import {Application} from "spectron"
import * as path from "path"

import {
  logIn,
  searchDisplay,
  setSpan,
  startSearch,
  waitForLoginAvailable,
  waitForHistogram,
  waitForSearch,
  writeSearch
} from "../lib/app.js"
import {TestTimeout, handleError} from "../lib/jest.js"
import {dataSets, selectors} from "../../src/js/test/integration"
import {retry} from "../lib/control"

describe("Correlation tests", () => {
  let app
  beforeEach(() => {
    app = new Application({
      path: electronPath,
      args: [path.join(__dirname, "..", "..")]
    })
    return app.start().then(() => app.webContents.send("resetState"))
  })

  afterEach(() => {
    if (app && app.isRunning()) {
      return app.stop()
    }
  })

  test(
    "conn for www.mybusinessdoc.com is found via correlation",
    (done) => {
      waitForLoginAvailable(app)
        .then(() => logIn(app))
        .then(() => waitForHistogram(app))
        .then(() => waitForSearch(app))
        .then(() => setSpan(app, dataSets.corelight.logDetails.span))
        .then(() =>
          writeSearch(app, dataSets.corelight.logDetails.initialSearch)
        )
        .then(() => startSearch(app))
        .then(() => waitForSearch(app))
        .then(() => searchDisplay(app))
        .then((results) => {
          expect(results).toMatchSnapshot()
        })
        .then(() =>
          app.client.rightClick(
            selectors.viewer.resultCellContaining(
              dataSets.corelight.logDetails.getDetailsFrom
            )
          )
        )
        .then(() =>
          app.client.click(selectors.viewer.rightClickMenuItem("Open details"))
        )
        .then(() =>
          Promise.all([
            retry(() => app.client.getText(selectors.correlationPanel.tsLabel)),
            retry(() => app.client.getText(selectors.correlationPanel.pathTag)),
            retry(() =>
              app.client.getText(selectors.correlationPanel.duration)
            ).then((result) => [result])
          ])
        )
        .then((correlationData) => {
          expect(correlationData).toMatchSnapshot()
          done()
        })
        .catch((err) => {
          handleError(app, err, done)
        })
    },
    TestTimeout
  )
})
