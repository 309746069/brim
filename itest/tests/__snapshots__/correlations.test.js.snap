// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Correlation tests conn for www.mybusinessdoc.com is found via correlation 1`] = `
Array [
  "ts",
  "_path",
  "uid",
  "id.orig_h",
  "id.orig_p",
  "id.resp_h",
  "id.resp_p",
  "trans_depth",
  "method",
  "host",
  "uri",
  "referrer",
  "version",
  "user_agent",
  "request_body_len",
  "response_body_len",
  "status_code",
  "status_msg",
  "info_code",
  "info_msg",
  "tags",
  "username",
  "password",
  "proxied",
  "orig_fuids",
  "orig_filenames",
  "orig_mime_types",
  "resp_fuids",
  "resp_filenames",
  "resp_mime_types",
  "04/07/15",
  "13:34",
  "48.1830",
  "http",
  "C9FG8S2NvxVUR0b0La",
  "192.168.0.53",
  "2210",
  "68.164.182.11",
  "80",
  "3",
  "GET",
  "www.mybusinessdoc.com",
  "/document.php?rnd=9813&id=5555525E01160D0F4A0C0E010809120D0F240309050D084A070B09",
  "-",
  "1.1",
  "Mozilla/4.0 (compatible; MSIE 7.0; Windows NT 5.1; Trident/4.0)",
  "0",
  "1,536",
  "200",
  "OK",
  "-",
  "-",
  "(empty)",
  "-",
  "-",
  "-",
  "-",
  "-",
  "-",
  "F2rjhJZAjwGdIvTM8",
  "f7.gif",
  "application/x-dosexec",
  "04/07/15",
  "13:34",
  "46.7830",
  "http",
  "C9FG8S2NvxVUR0b0La",
  "192.168.0.53",
  "2210",
  "68.164.182.11",
  "80",
  "2",
  "GET",
  "www.mybusinessdoc.com",
  "/document.php?rnd=5292&id=5555525E01160D0F4A0C0E010809120D0F240309050D084A070B09",
  "-",
  "1.1",
  "Mozilla/4.0 (compatible; MSIE 7.0; Windows NT 5.1; Trident/4.0)",
  "0",
  "192,512",
  "200",
  "OK",
  "-",
  "-",
  "(empty)",
  "-",
  "-",
  "-",
  "-",
  "-",
  "-",
  "FuFeW33aTSTxXihCql",
  "551d88323f7e.gif",
  "application/x-dosexec",
  "04/07/15",
  "13:34",
  "44.2440",
  "http",
  "C9FG8S2NvxVUR0b0La",
  "192.168.0.53",
  "2210",
  "68.164.182.11",
  "80",
  "1",
  "GET",
  "www.mybusinessdoc.com",
  "/document.php?rnd=3271&id=5555525E01160D0F4A0C0E010809120D0F240309050D084A070B09",
  "-",
  "1.1",
  "Mozilla/4.0 (compatible; MSIE 7.0; Windows NT 5.1; Trident/4.0)",
  "0",
  "318,883",
  "200",
  "OK",
  "-",
  "-",
  "(empty)",
  "-",
  "-",
  "-",
  "-",
  "-",
  "-",
  "FGx5ts2iCMfZSUgO8c",
  "c87ed3c.gif",
  "application/x-dosexec",
]
`;

exports[`Correlation tests conn for www.mybusinessdoc.com is found via correlation 2`] = `
Array [
  Array [
    "13:34:43.772",
    "13:34:44.244",
    "13:34:44.563",
    "13:34:46.783",
    "13:34:47.442",
    "13:34:48.183",
    "13:34:48.373",
  ],
  Array [
    "conn",
    "http",
    "files",
    "http",
    "files",
    "http",
    "files",
  ],
  Array [
    "Duration: 4.601s",
  ],
]
`;
